### ISSUE TITLE
list_url not working

### ISSUE THREAD

**Issue by glennpj** [LLNL/spack#1225]
One of the packages I was trying to install as a dependency has been updated. There is a `list_url` in the package file but it does not get checked when fetching.
```
   homepage = "https://github.com/jeroenooms/jsonlite"                                                                 
    url      = "https://cran.r-project.org/src/contrib/jsonlite_0.9.21.tar.gz"                                          
    list_url = "https://cran.r-project.org/src/contrib/Archive/jsonlite"                                                
                                                                                                                        
    version('0.9.21', '4fc382747f88a79ff0718a0d06bed45d') 
```
```
spack fetch r-jsonlite
==> Trying to fetch from file:///home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-0.9.21.tar.gz
curl: (37) Couldn't open file /home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-0.9.21.tar.gz
==> Fetching from file:///home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-0.9.21.tar.gz failed.
==> Trying to fetch from https://cran.r-project.org/src/contrib/jsonlite_0.9.21.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from https://cran.r-project.org/src/contrib/jsonlite_0.9.21.tar.gz failed.
==> Error: All fetchers failed for r-jsonlite-0.9.21-oegoqbzsaflpwyndb5qs5q7dhwayjxw3
```

The file exists at the URL specified by `list_url`.
https://cran.r-project.org/src/contrib/Archive/jsonlite/
https://cran.r-project.org/src/contrib/Archive/jsonlite/jsonlite_0.9.21.tar.gz

Any ideas?

Thanks.

**Comment by adamjstewart** [comment]
It looks like `list_url` is being overridden by `url`. Just get rid of `list_url` and replace `url` with the correct URL.

**Comment by adamjstewart** [comment]
To be honest, I'm not sure if `list_url` has ever worked for me...

**Comment by glennpj** [comment]
But it needs `list_url` to keep track of the older versions per the [documentation](http://software.llnl.gov/spack/packaging_guide.html?highlight=list_url#list-url).

**Comment by adamjstewart** [comment]
Ah, I see. Let me poke around and see what I can find. `spack checksum` never works for me either whenever I need to use `list_url`. I've been meaning to take a look at that.

If you just want to get things working for now, you can define a `url_for_version` function (see mvapich2 for an example).

**Comment by glennpj** [comment]
Yea, I can work around it but I wanted to report the issue because I see the potential for a whole bunch of packages breaking when updated upstream.

Thanks.

**Comment by adamjstewart** [comment]
I took a look, but I think I'm out of my element. There's fetch stuff spread across several files. Hopefully someone else can help.

**Comment by glennpj** [comment]
I am starting to wonder if maybe I have misunderstood what `list_url` is for. Is it just for finding versions via `spack checksum`? I was under the impression that it would be used anytime spack was looking for "available versions of packages" as stated in the docs.

@tgamblin 
Could you chime in on this? Have I discovered a misbehavior in spack or have I misunderstood the list_url feature?

**Comment by adamjstewart** [comment]
It wouldn't really be useful if it was only for finding versions. Also, it can't even do that!:
```python
class Mpich(Package):                                                           
    """MPICH is a high performance and widely portable implementation of        
    the Message Passing Interface (MPI) standard."""                         
    homepage = "http://www.mpich.org"                                           
    url      = "http://www.mpich.org/static/downloads/3.0.4/mpich-3.0.4.tar.gz"      
    list_url = "http://www.mpich.org/static/downloads/"                         
    list_depth = 2
```
$ spack versions mpich
==> Safe versions (already checksummed):
  3.2  3.1.4  3.1.3  3.1.2  3.1.1  3.1  3.0.4
==> Remote versions (not yet checksummed):
  Found no unckecksummed versions for mpich
```

**Comment by tobbez** [comment]
`spack.util.web._spider` seems to be subtly broken.

```
$ spack python
>>> import spack.util.web
>>> pages, links = spack.util.web._spider(('https://mesa.freedesktop.org/archive/', set(), 'https://mesa.freedesktop.org/archive/', None, 1, 3, True))
>>> pages.keys() 
['https://mesa.freedesktop.org/archive/?C=N;O=D', 'https://mesa.freedesktop.org/archive/']
>>> import llnl.util.tty
>>> llnl.util.tty.set_debug(True)
>>> pages, links = spack.util.web._spider(('https://mesa.freedesktop.org/archive/', set(), 'https://mesa.freedesktop.org/archive/', None, 1, 3, True))
==> Error in _spider: local variable 'pool' referenced before assignment

# Edit lib/spack/spack/util/web.py and add "raise" at the end of the except clause for the try clause wrapping the whole function, then:
>>> reload(spack.util.web)
>>> pages, links = spack.util.web._spider(('https://mesa.freedesktop.org/archive/', set(), 'https://mesa.freedesktop.org/archive/', None, 1, 3, True))
==> Error in _spider: local variable 'pool' referenced before assignment
==> Error in _spider: local variable 'pool' referenced before assignment
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/ketl/spack/lib/spack/spack/util/web.py", line 137, in _spider
    results = pool.map(_spider, subcalls)
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 251, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
UnboundLocalError: local variable 'pool' referenced before assignment
```

On a side note, the name of the parameter `raise_on_error` is a bit misleading, since it only has any effect for `urllib2.URLError`. Any other exceptions are silently ignored unless debug mode is turned on.

**Comment by tobbez** [comment]
After moving the initialization of the pool in `_spider` (`pool = Pool(processes=len(subcalls))`) out of the try block it resides in, placing it before it, you get:
```
>>> pages, links = spack.util.web._spider(('https://mesa.freedesktop.org/archive/', set(), 'https://mesa.freedesktop.org/archive/', None, 1, 3, True))
==> Error in _spider: daemonic processes are not allowed to have children
==> Error in _spider: daemonic processes are not allowed to have children
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/ketl/spack/lib/spack/spack/util/web.py", line 138, in _spider
    results = pool.map(_spider, subcalls)
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 251, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
AssertionError: daemonic processes are not allowed to have children
```

So it would seem recursing within _spider using multiprocessing is not allowed.

That means `_spider` likely requires some rewriting.

A couple of observations:
* `list_depth` likely hasn't seen much use (or users just haven't noticed or cared enough to report it earlier)
* Even if the existing code worked, it could potentially cause a fork bomb (when `maxdepth`>2, and there are many links on all levels of the hierarchy).

**Comment by glennpj** [comment]
Well, `spack versions` actually does work.
```
spack versions r-jsonlite
==> Safe versions (already checksummed):
  0.9.21
==> Remote versions (not yet checksummed):
  1.0     0.9.20  0.9.18  0.9.16  0.9.14  0.9.12  0.9.10  0.9.8  0.9.6  0.9.4  0.9.1
  0.9.22  0.9.19  0.9.17  0.9.15  0.9.13  0.9.11  0.9.9   0.9.7  0.9.5  0.9.3  0.9.0
```
But when fetching it only looks at the `url` and not the `list_url`.

**Comment by glennpj** [comment]
So, this problem is even worse than I thought. If I add a `url` attribute to `version` to get r-jsonlite-0.9.21 to fetch properly, then updating to the new version is broken because the fetcher looks at the wrong URL.
```
   homepage = "https://github.com/jeroenooms/jsonlite"                                                                 
    url      = "https://cran.r-project.org/src/contrib/jsonlite_0.9.21.tar.gz"                                          
    list_url = "https://cran.r-project.org/src/contrib/Archive/jsonlite"                                                
                                                                                                                        
    version('1.0', 'c8524e086de22ab39b8ac8000220cc87')                                                                  
    version('0.9.21', '4fc382747f88a79ff0718a0d06bed45d',                                                               
            url='https://cran.r-project.org/src/contrib/Archive/jsonlite/jsonlite_0.9.21.tar.gz')
```
```
spack fetch r-jsonlite@0.9.21
==> Trying to fetch from file:///home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-0.9.21.tar.gz
curl: (37) Couldn't open file /home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-0.9.21.tar.gz
==> Fetching from file:///home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-0.9.21.tar.gz failed.
==> Trying to fetch from https://cran.r-project.org/src/contrib/Archive/jsonlite/jsonlite_0.9.21.tar.gz
######################################################################## 100.0%
```
But now the new version can not be fetched.
```
spack fetch r-jsonlite@1.0
==> Trying to fetch from file:///home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-1.0.tar.gz
curl: (37) Couldn't open file /home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-1.0.tar.gz
==> Fetching from file:///home/gjohnson/spack/var/spack/cache/r-jsonlite/r-jsonlite-1.0.tar.gz failed.
==> Trying to fetch from https://cran.r-project.org/src/contrib/Archive/jsonlite/jsonlite_1.0.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from https://cran.r-project.org/src/contrib/Archive/jsonlite/jsonlite_1.0.tar.gz failed.
==> Error: All fetchers failed for r-jsonlite-1.0-c2lzktjjz6wwzo674poocphmktxw3d2y
```

**Comment by glennpj** [comment]
@tgamblin 
Could you take a look at this issue? It is kind of a show stopper for R packages but may affect other packages as well.

Thanks.

**Comment by glennpj** [comment]
After looking through the code and doing some empirical tests, I have concluded that `list_url` is not what I thought it was. What is really needed for older tar files that get moved to an archive on the remote site is having the `url` property return a list as outlined in #1090.

**Comment by adamjstewart** [comment]
So... what is `list_url` for?

**Comment by glennpj** [comment]
Since I started an attempt to make `list_url` usable for fetching in #1346 I am reopening this issue.

**Comment by glennpj** [comment]
This can be closed now that #1346 has been merged.
